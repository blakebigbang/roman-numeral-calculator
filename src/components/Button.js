import styled from "@emotion/styled";

import UserInterface from "../utils/UserInterface";

const Button = styled.button`
border: none;
background: none;
color: #fff;
user-select: none;
&:focus {
  outline: none;
}
display: flex;
justify-content: center;
align-items: center;

position: ${props => {
  if (props.buttonType === UserInterface.BUTTON_TYPES.OPERATION) {
    return `relative`;
  }
}};

/* top: ${props => {
  if (props.buttonType === UserInterface.BUTTON_TYPES.OPERATION) {
    return `0.05rem`;
  }
}}; */

font-size: ${props => {
  if (props.buttonType === UserInterface.BUTTON_TYPES.OPERATION) {
    return `4rem`;
  }

  return `4rem`;
}};

${UserInterface.MEDIA_QUERIES[0]} {
  font-size: ${props => {
    if (props.buttonType === UserInterface.BUTTON_TYPES.OPERATION) {
      return `calc(4rem / 2)`;
    }

    return `calc(4rem / 2)`;
  }};
}

${UserInterface.MEDIA_QUERIES[1]} {
  font-size: ${props => {
    if (props.buttonType === UserInterface.BUTTON_TYPES.OPERATION) {
      return `calc(4rem / 3)`;
    }

    return `calc(4rem / 3)`;
  }};
}

border: ${props => {
  if (props.active) {
    return `3px solid black`;
  }

  return `3px solid transparent`;
}};

background-color: ${props => {
  if (props.buttonType === UserInterface.BUTTON_TYPES.EDIT) {
    return `#414246`;
  } else if (props.buttonType === UserInterface.BUTTON_TYPES.OPERATION) {
    return `#FF9F0A`;
  } else if (props.buttonType === UserInterface.BUTTON_TYPES.NUMERAL) {
    return `#5D6365`;
  }
}};

&:active,
&:focus {
  background-color: ${props => {
    if (props.buttonType === UserInterface.BUTTON_TYPES.EDIT) {
      return `#5D6365`;
    } else if (props.buttonType === UserInterface.BUTTON_TYPES.OPERATION) {
      return `#CD7D03`;
    } else if (props.buttonType === UserInterface.BUTTON_TYPES.NUMERAL) {
      return `#A0A1A4`;
    }
  }};
}
`;

export default Button;
