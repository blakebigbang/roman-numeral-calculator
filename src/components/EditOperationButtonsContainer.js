import styled from "@emotion/styled";

const EditOperationButtonsContainer = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-template-rows: 100%;
  grid-column-gap: 0.15rem;
  /* grid-row-gap: 0.15rem; */
  padding-bottom: 0.15rem;
`;

export default EditOperationButtonsContainer;
