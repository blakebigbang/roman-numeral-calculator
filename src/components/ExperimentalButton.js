import styled from "@emotion/styled";

import UI from "../utils/UI";

const ExperimentalButton = styled.button`
  background-color: ${props => {
    if (props.active && props.buttonType === UI.BUTTON_TYPES.OPERATION) {
      return `#00ff74 !important`;
    }
  }};

  color: ${props => {
    if (props.active && props.buttonType === UI.BUTTON_TYPES.OPERATION) {
      return `#131919 !important`;
    }
  }};

  &:active,
  &:focus {
    background-color: ${props => {
      if (props.buttonType === UI.BUTTON_TYPES.EDIT) {
        return `#ffc000`;
      } else if (props.buttonType === UI.BUTTON_TYPES.OPERATION) {
        return `#00ff74`;
      } else if (props.buttonType === UI.BUTTON_TYPES.NUMERAL) {
        return `#00c8ff`;
      }
    }};
    color: ${props => {
      if (props.buttonType === UI.BUTTON_TYPES.EDIT) {
        return `#131919`;
      } else if (props.buttonType === UI.BUTTON_TYPES.OPERATION) {
        return `#131919`;
      } else if (props.buttonType === UI.BUTTON_TYPES.NUMERAL) {
        return `#131919`;
      }
    }};
  }
`;

export default ExperimentalButton;
