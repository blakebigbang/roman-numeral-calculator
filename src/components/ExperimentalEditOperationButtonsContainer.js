import styled from "@emotion/styled";

const ExperimentalEditOperationButtonsContainer = styled.div`
  position: absolute;
  top: 0;
  width: 100%;
  padding: 1rem;

  display: flex;
  justify-content: space-between;

  button {
    background: rgba(255, 159, 10, 0.3);
    border: 1px solid #ffc000;

    font-family: "IBM Plex Mono";
    font-size: 20px;
    color: #ffc000;
    border-radius: 50%;
    width: 3rem;
    height: 3rem;
    display: flex;
    justify-content: center;
    align-items: center;
  }
`;

export default ExperimentalEditOperationButtonsContainer;
