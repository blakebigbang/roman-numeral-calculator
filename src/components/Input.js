import styled from "@emotion/styled";
import UserInterface from "../utils/UserInterface";

const Input = styled.input`
  display: block;
  width: 100%;
  background-color: #2a3131;
  border: none;
  color: #fff;
  font-size: 5rem;
  padding: 2rem;
  padding-top: 4rem;

  ${UserInterface.MEDIA_QUERIES[0]} {
    font-size: calc(5rem / 1);
    padding: calc(2rem / 1);
    padding-top: calc(4rem / 1);
  }

  ${UserInterface.MEDIA_QUERIES[1]} {
    font-size: calc(5rem / 1.4);
    padding: calc(2rem / 1.4);
    padding-top: calc(4rem / 1.4);
  }

  outline: none;
  text-align: right;
  border-radius: 0;
  -webkit-appearance: none;
`;

export default Input;
