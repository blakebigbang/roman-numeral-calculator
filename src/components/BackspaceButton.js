import React from "react";

import Button from "./Button";
import ExperimentalButton from "./ExperimentalButton";

import THEMES_KEYS from "../data/THEMES_KEYS";
import BUTTON_TYPES from "../data/BUTTON_TYPES";

function BackspaceButton({
  theme,
  operationMode,
  first,
  second,
  vibrate,
  setFirst,
  setSecond
}) {
  const showBackspace =
    (!operationMode && first.length > 0) ||
    (operationMode && second.length > 0);

  if (theme === THEMES_KEYS.CLASSIC) {
    return (
      <Button
        buttonType={BUTTON_TYPES.EDIT}
        tabIndex={1}
        disabled={!showBackspace}
        style={{ opacity: showBackspace ? 1 : 0.4 }}
        onClick={() => {
          vibrate();

          if (!operationMode) {
            return setFirst(first.substr(0, first.length - 1));
          }

          return setSecond(second.substr(0, second.length - 1));
        }}
      >
        &larr;
      </Button>
    );
  } else if (theme === THEMES_KEYS.EXPERIMENTAL) {
    return (
      <ExperimentalButton
        buttonType={BUTTON_TYPES.EDIT}
        tabIndex={1}
        disabled={!showBackspace}
        style={{ opacity: showBackspace ? 1 : 0.4 }}
        onClick={() => {
          vibrate();

          if (!operationMode) {
            return setFirst(first.substr(0, first.length - 1));
          }

          return setSecond(second.substr(0, second.length - 1));
        }}
      >
        &larr;
      </ExperimentalButton>
    );
  }
}

export default BackspaceButton;
