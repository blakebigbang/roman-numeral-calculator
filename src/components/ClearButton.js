import React from "react";

import Button from "./Button";
import ExperimentalButton from "./ExperimentalButton";

import UI from "../utils/UI";

export const handleClear = ({
  vibrate,
  setError,
  first,
  second,
  setPreviousOperation,
  operationMode,
  setFirst,
  setOperationMode,
  setSecond
}) => {
  vibrate();
  setError(false);

  if (!first && !second) {
    return setPreviousOperation();
  } else if (first && !operationMode && !second) {
    return setFirst("");
  } else if (first && operationMode && !second) {
    return setOperationMode();
  } else if (first && !operationMode && second) {
    // impossible
  } else if (first && operationMode && second) {
    return setSecond("");
  }
};

function ClearButton({
  theme,
  vibrate,
  setError,
  first,
  second,
  setPreviousOperation,
  operationMode,
  setFirst,
  setOperationMode,
  setSecond
}) {
  if (theme === UI.THEMES_KEYS.CLASSIC) {
    return (
      <Button
        buttonType={UI.BUTTON_TYPES.EDIT}
        tabIndex={0}
        onClick={() =>
          handleClear({
            vibrate,
            setError,
            first,
            second,
            setPreviousOperation,
            operationMode,
            setFirst,
            setOperationMode,
            setSecond
          })
        }
      >
        AC
      </Button>
    );
  } else if (theme === UI.THEMES_KEYS.EXPERIMENTAL) {
    return (
      <ExperimentalButton
        buttonType={UI.BUTTON_TYPES.EDIT}
        tabIndex={0}
        onClick={() =>
          handleClear({
            vibrate,
            setError,
            first,
            second,
            setPreviousOperation,
            operationMode,
            setFirst,
            setOperationMode,
            setSecond
          })
        }
      >
        AC
      </ExperimentalButton>
    );
  }
  return null;
}

export default ClearButton;
