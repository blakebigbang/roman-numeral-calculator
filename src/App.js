import React, { useState } from "react";

import ThemeBar from "./components/ThemeBar";
import ClearButton from "./components/ClearButton";
import BackspaceButton from "./components/BackspaceButton";
import Input from "./components/Input";
import NumeralButton from "./components/NumeralButton";
import OperationButton from "./components/OperationButton";
import ExperimentalInputCircleContainer from "./components/ExperimentalInputCircleContainer";
import RootContainer from "./components/RootContainer";
import RootButtonsContainer from "./components/RootButtonsContainer";
import ExperimentalInnerContainer from "./components/ExperimentalInnerContainer";
import MainButtonsContainer from "./components/MainButtonsContainer";
import EditOperationButtonsContainer from "./components/EditOperationButtonsContainer";
import NumeralButtonsContainer from "./components/NumeralButtonsContainer";
import OperationButtonsContainer from "./components/OperationButtonsContainer";
import ExperimentalOperationButtonsContainer from "./components/ExperimentalOperationButtonsContainer";
import ExperimentalEditOperationButtonsContainer from "./components/ExperimentalEditOperationButtonsContainer";
import ExperimentalNumeralButtonsContainer from "./components/ExperimentalNumeralButtonsContainer";
import ExperimentalNumeralRow from "./components/ExperimentalNumeralRow";
import ExperimentalInputContainer from "./components/ExperimentalInputContainer";
import ExperimentalInput from "./components/ExperimentalInput";

import NUMERALS from "./data/NUMERALS";
import OPERATIONS, { OPERATIONS_KEYS } from "./data/OPERATIONS";
import UI from "./utils/UI";
import convert from "./utils/convert";

function App(props) {
  const [first, setFirst] = useState(props.first || "");
  const [second, setSecond] = useState(props.second || "");
  const [operationMode, setOperationMode] = useState(
    props.operationMode || undefined
  );
  const [previousOperation, setPreviousOperation] = useState();
  const [hasError, setError] = useState(false);
  const [theme, setTheme] = useState(UI.THEMES_KEYS.EXPERIMENTAL);
  const { width } = UI.useWindowSize();

  // EXPERIMENTAL THEME
  if (theme === UI.THEMES_KEYS.EXPERIMENTAL) {
    return (
      <RootContainer>
        <ExperimentalInnerContainer>
          <ExperimentalInputContainer>
            <ExperimentalInputCircleContainer
              width={width}
              hasError={hasError}
              previousOperation={previousOperation}
            />
            {!operationMode || operationMode === OPERATIONS_KEYS.EQUALS ? (
              <ExperimentalInput data-testid="input" value={first} readOnly />
            ) : (
              <ExperimentalInput data-testid="input" value={second} readOnly />
            )}
          </ExperimentalInputContainer>

          <ExperimentalEditOperationButtonsContainer>
            <ClearButton
              theme={theme}
              vibrate={UI.vibrate}
              setError={setError}
              first={first}
              second={second}
              setPreviousOperation={setPreviousOperation}
              operationMode={operationMode}
              setFirst={setFirst}
              setOperationMode={setOperationMode}
              setSecond={setSecond}
            />
            <BackspaceButton
              theme={theme}
              operationMode={operationMode}
              first={first}
              second={second}
              vibrate={UI.vibrate}
              setFirst={setFirst}
              setSecond={setSecond}
            />
          </ExperimentalEditOperationButtonsContainer>

          <ExperimentalNumeralButtonsContainer>
            <ExperimentalNumeralRow>
              {["I", "V", "X"].map((numeral, index) => (
                <NumeralButton
                  key={numeral}
                  theme={theme}
                  numeral={numeral}
                  index={index}
                  vibrate={UI.vibrate}
                  operationMode={operationMode}
                  setFirst={setFirst}
                  first={first}
                  setSecond={setSecond}
                  second={second}
                />
              ))}
            </ExperimentalNumeralRow>
            <ExperimentalNumeralRow>
              {["L", "C", "D", "M"].map((numeral, index) => (
                <NumeralButton
                  key={numeral}
                  theme={theme}
                  numeral={numeral}
                  index={index}
                  vibrate={UI.vibrate}
                  operationMode={operationMode}
                  setFirst={setFirst}
                  first={first}
                  setSecond={setSecond}
                  second={second}
                />
              ))}
            </ExperimentalNumeralRow>
          </ExperimentalNumeralButtonsContainer>

          <ExperimentalOperationButtonsContainer>
            {Object.keys(OPERATIONS).map((opKey, index) => (
              <OperationButton
                key={opKey}
                theme={theme}
                index={index}
                opKey={opKey}
                operationMode={operationMode}
                vibrate={UI.vibrate}
                first={first}
                convert={convert}
                setError={setError}
                previousOperation={previousOperation}
                setPreviousOperation={setPreviousOperation}
                setOperationMode={setOperationMode}
                setFirst={setFirst}
                second={second}
                setSecond={setSecond}
              />
            ))}
          </ExperimentalOperationButtonsContainer>
          <ThemeBar theme={theme} setTheme={setTheme} />
        </ExperimentalInnerContainer>
      </RootContainer>
    );
  }

  // CLASSIC THEME
  return (
    <RootContainer>
      {!operationMode || operationMode === OPERATIONS_KEYS.EQUALS ? (
        <Input data-testid="input" value={first} readOnly />
      ) : (
        <Input data-testid="input" value={second} readOnly />
      )}

      <RootButtonsContainer>
        <MainButtonsContainer>
          <EditOperationButtonsContainer>
            <ClearButton
              theme={theme}
              vibrate={UI.vibrate}
              setError={setError}
              first={first}
              second={second}
              setPreviousOperation={setPreviousOperation}
              operationMode={operationMode}
              setFirst={setFirst}
              setOperationMode={setOperationMode}
              setSecond={setSecond}
            />
            <BackspaceButton
              theme={theme}
              operationMode={operationMode}
              first={first}
              second={second}
              vibrate={UI.vibrate}
              setFirst={setFirst}
              setSecond={setSecond}
            />
          </EditOperationButtonsContainer>

          <NumeralButtonsContainer>
            {Object.keys(NUMERALS).map((numeral, index) => (
              <NumeralButton
                key={numeral}
                theme={theme}
                numeral={numeral}
                index={index}
                vibrate={UI.vibrate}
                operationMode={operationMode}
                setFirst={setFirst}
                first={first}
                setSecond={setSecond}
                second={second}
              />
            ))}
          </NumeralButtonsContainer>
        </MainButtonsContainer>

        <OperationButtonsContainer>
          {Object.keys(OPERATIONS).map((opKey, index) => (
            <OperationButton
              key={opKey}
              theme={theme}
              index={index}
              opKey={opKey}
              operationMode={operationMode}
              vibrate={UI.vibrate}
              first={first}
              convert={convert}
              setError={setError}
              previousOperation={previousOperation}
              setPreviousOperation={setPreviousOperation}
              setOperationMode={setOperationMode}
              setFirst={setFirst}
              second={second}
              setSecond={setSecond}
            />
          ))}
        </OperationButtonsContainer>
      </RootButtonsContainer>
      <ThemeBar theme={theme} setTheme={setTheme} />
    </RootContainer>
  );
}

export default App;
