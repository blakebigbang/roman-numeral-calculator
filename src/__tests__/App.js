import React from "react";
import renderer from "react-test-renderer";
import { shallow, mount, render } from "enzyme";

import App from "../App";
import Input from "../components/Input";
import { handleClear } from "../components/ClearButton";

describe("App.js", () => {
  test("renders", () => {
    const component = renderer.create(<App />);

    let tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  test("handleClear clears text", () => {
    // const wrapper = render(<App />);
    // console.log(wrapper);
    // const input = wrapper.find("input");
    // input.simulate("focus");
    // input.simulate("change", { target: { value: "XX" } });
    // expect(input.get(0).value).toBe("XX");
    // expect(input.props()).toEqual("X");
  });
});
