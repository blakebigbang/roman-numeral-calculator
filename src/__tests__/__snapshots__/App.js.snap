// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.js renders 1`] = `
<div
  className="css-18nhuy4"
>
  <div
    className="css-ko3v73"
  >
    <div
      className="css-2cdian"
    >
      <svg
        height={300}
        width={1024}
      >
        <circle
          cx={512}
          cy={150}
          fill="transparent"
          r={120}
          stroke="white"
          strokeDasharray="1,6"
          strokeWidth={1}
        >
          <animate
            attributeName="stroke-width"
            attributeType="XML"
            dur="4s"
            from={1}
            repeatCount="indefinite"
            to={10}
          />
        </circle>
        <circle
          cx={512}
          cy={150}
          fill="transparent"
          opacity={0.1}
          r={103.44827586206897}
          stroke="white"
          strokeWidth={1}
        >
          <animate
            attributeName="opacity"
            attributeType="XML"
            dur="4s"
            from={0}
            repeatCount="indefinite"
            to={0.3}
          />
        </circle>
      </svg>
      <input
        className="css-1kj977d"
        data-testid="input"
        readOnly={true}
        value=""
      />
    </div>
    <div
      className="css-j6x9og"
    >
      <button
        className="css-kndt0l"
        onClick={[Function]}
        tabIndex={0}
      >
        AC
      </button>
      <button
        className="css-kndt0l"
        disabled={true}
        onClick={[Function]}
        style={
          Object {
            "opacity": 0.4,
          }
        }
        tabIndex={1}
      >
        ←
      </button>
    </div>
    <div
      className="css-rlf52g"
    >
      <div
        className="css-ov1ktg"
      >
        <button
          className="css-45xxc6"
          id="I"
          onClick={[Function]}
          tabIndex={2}
        >
          I
        </button>
        <button
          className="css-45xxc6"
          id="V"
          onClick={[Function]}
          tabIndex={3}
        >
          V
        </button>
        <button
          className="css-45xxc6"
          id="X"
          onClick={[Function]}
          tabIndex={4}
        >
          X
        </button>
      </div>
      <div
        className="css-ov1ktg"
      >
        <button
          className="css-45xxc6"
          id="L"
          onClick={[Function]}
          tabIndex={2}
        >
          L
        </button>
        <button
          className="css-45xxc6"
          id="C"
          onClick={[Function]}
          tabIndex={3}
        >
          C
        </button>
        <button
          className="css-45xxc6"
          id="D"
          onClick={[Function]}
          tabIndex={4}
        >
          D
        </button>
        <button
          className="css-45xxc6"
          id="M"
          onClick={[Function]}
          tabIndex={5}
        >
          M
        </button>
      </div>
    </div>
    <div
      className="css-1vupdkq"
    >
      <button
        className="css-dyzt8p"
        onClick={[Function]}
        tabIndex={9}
      >
        <span>
          +
        </span>
      </button>
      <button
        className="css-dyzt8p"
        onClick={[Function]}
        tabIndex={10}
      >
        <span>
          −
        </span>
      </button>
      <button
        className="css-dyzt8p"
        onClick={[Function]}
        tabIndex={11}
      >
        <span>
          ×
        </span>
      </button>
      <button
        className="css-dyzt8p"
        onClick={[Function]}
        tabIndex={12}
      >
        <span>
          ÷
        </span>
      </button>
      <button
        className="css-dyzt8p"
        onClick={[Function]}
        tabIndex={13}
      >
        <span>
          =
        </span>
      </button>
    </div>
    <div
      className="css-1n22ovz"
    >
      <button
        className="css-1mmka7t"
        onClick={[Function]}
      >
        Classic
      </button>
      <button
        className="css-47h7xl"
        onClick={[Function]}
      >
        Experimental
      </button>
    </div>
  </div>
</div>
`;
